fuel-tx/src/transaction/types/storage.rs

```diff
@@ -16,51 +15,157 @@ use rand::{
+    fn from(c: Vec<u8>) -> Self {
+        Self(c)
+    }
+}
+
+impl From<&[u8]> for ContractsStateData {
+    fn from(c: &[u8]) -> Self {
+        Self(c.into())
+    }
+}
+
+impl From<&mut [u8]> for ContractsStateData {
+    fn from(c: &mut [u8]) -> Self {
+        Self(c.into())
+    }
+}
+
+impl From<ContractsStateData> for Vec<u8> {
+    fn from(c: ContractsStateData) -> Vec<u8> {
+        c.0
+    }
+}
+
+impl AsRef<[u8]> for ContractsStateData {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl AsMut<[u8]> for ContractsStateData {
+    fn as_mut(&mut self) -> &mut [u8] {
+        self.0.as_mut()
+    }
+}
+
+#[cfg(feature = "random")]
+impl Distribution<ContractsStateData> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> ContractsStateData {
+        ContractsStateData(rng.gen::<Bytes32>().to_vec())
+    }
+}
+
+impl IntoIterator for ContractsStateData {
+    type IntoIter = alloc::vec::IntoIter<Self::Item>;
+    type Item = u8;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+// TODO: Remove manual serialization when implementing dynamic storage
+// StorageData uses Vec<u8> to represent storage data. We manually implement
+// Serialize and Deserialize in order to preserve the 32-byte serialized format
+// of the StorageData, and by extension, the Create Transaction. When it is
+// possible to write dynamically sized storage data via new opcodes, and
+// dynamically sized slots are supported, remove these implementations, allowing
+// native support for dynamically sized serialization and deserialization.
+impl Serialize for ContractsStateData {
+    fn size_static(&self) -> usize {
+        Bytes32::LEN
+    }
+
+    fn size_dynamic(&self) -> usize {
+        0
+    }
+
+    fn encode_static<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        let mut bytes = Bytes32::default();
+        bytes[0..Bytes32::LEN].copy_from_slice(&self.0[0..Bytes32::LEN]);
+        bytes.encode_static(buffer)
+    }
+
+    fn encode_dynamic<O: Output + ?Sized>(&self, _buffer: &mut O) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+// TODO: Remove manual deserialization when implementing dynamic storage
+impl Deserialize for ContractsStateData {
+    fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+        let bytes = Bytes32::decode(buffer)?;
+        let data = Self::from(bytes.as_ref());
+        Ok(data)
+    }
+
+    fn decode_dynamic<I: Input + ?Sized>(
+        &mut self,
+        _buffer: &mut I,
+    ) -> Result<(), Error> {
+        Ok(())
+    }
+}
 
 #[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 #[cfg_attr(feature = "typescript", wasm_bindgen::prelude::wasm_bindgen)]
 #[derive(Deserialize, Serialize)]
 pub struct StorageSlot {
     key: Bytes32,
-    value: Bytes32,
+    value: ContractsStateData,
```

Comment:
Let's not change the type for the `Transaction` to avoid breaking changes and to not affect the serialziation